### DISCUSSION TITLE
More control on Gameplay functions in PGNViewer

### DISCUSSION THREAD

**Discussion by coffeeDev98** [mliebelt/pgn-viewer#319] [Q&A]
Just curious, Is there any reason why you didn't get into gameplay for PGNViewer. Like, chessground's functionalities... to control what to do when a player moves a piece, etc...
Would like an opportunity to discuss this with you... And, also, the possibility of implementing this in the future...

**Comment by mliebelt** [comment]
Thank you a lot for your question / recommendation, even if I don't have an answer yet. Let me rephrase what was happening in the past.

I started with my own implementation on base of chessboard.js and chess.js, which went smooth. Sometimes later, I saw Lichess.org grow, and saw the incredible chessboard there used, named Chessground. And since that, I was keen to use Chessground _as it is_, because I don't think I can improve it in any way. Small parts of it I configured, and I found hacks to style what I don't liked, but most of it is perfect in how it is.

What are the parts of Chessground you don't like, or where you think could be improved? I don't have any idea ...

**Reply by coffeeDev98** [reply]
Yes, agree on chessground being the ideal package for chessboards... But won't it be possible to cover more use-cases if the functions that handle what happens when the user plays a move are exposed outside(maybe a separate mode in itself - view, print, edit, **`play`**). This can give more flexibility to whoever is using this package.  

What do you think about the scope of this...?

**Reply by mliebelt** [reply]
This sound similar to something I want to do a long time ... See https://github.com/mliebelt/PgnViewerJS/issues/175 and https://github.com/mliebelt/PgnViewerJS/issues/266 for tickets that correspond to it. I think that defining an API, that can then be used by others would be the right way to go.

What I am currently doing (by moving the reader (first) to Typescript) is to make the existing API more visible, and then I want to see if the reader provides the right API for all use cases. I would like to refactor parts of it, so to have then finally the "right" API from the reader.

Second step will be then to provide one (or many?) APIs for the viewer. This could depend on the mode you are in, so perhaps we will have at the end a Player (for playing through already existing games), an Editor to adjust games in all possible ways, ...
